### PR TITLE
Refactored the gene-go annotation functions

### PR DESCRIPTION
**Pull request by Habush**
- Gene-GO annotation functions don't use global variables
- The parents are fetched in tail-recursive manner 
- Removed unnecessary definition in `utils.scm`
- `gene-go-annotation` function returns a `hash-table`, where gene names are keys and the go terms are values,  instead of a `list`

### CODE REVIEW

**Comment by tanksha** [comment]
I don't think returning a hash-table is helpful, two of my reasons are

- The parser don't depend on the genes, it only looks for OpenCog links (EvaluationLink, MemberLink or InheritanceLink) and parse the atoms as nodes and relationship as edges accordingly.

-The scm result of the annotations should be importable into the atomspace for later uses, changes we are making for the parser, should not affect the result.

- Inorder to have importable result, we will use ListLink other than a list hence, the parser will have access to the elements of the ListLink as a list by using opencog util functions like (cog-outgoing-set )

**Comment by Habush** [comment]
@tanksha I have refactored the `gene-go-annotation` to return `ListLink` per your comment. Check it out.

File: functions/gene_go_annotation.scm [hunk 1]
```diff
@@ -1,11 +1,9 @@
 (define* (gene-go-annotation gene_nodes namespace #:optional (parents 0))
     (let (
-      [hs (make-hash-table)]
     )
-  (for-each (lambda (gene) 
-      (hash-set! hs gene (find-go-term gene namespace parents))
+  (ListLink (flatten (map (lambda (gene)
```
Comment (tanksha) [inline functions/gene_go_annotation.scm]: @Habush I found empty result testing it. I think the reason is result from (find-go-term) for each gene are not collected.  What I recommend will be:

(define* (gene-go-annotation gene_nodes namespace #:optional (parents 0))
  (let ([result '()])
  (for-each (lambda (gene)
  (set! result (append result (find-go-term gene namespace parents)))
  )gene_nodes)

 (ListLink result)
)) 

As the output of  (find-go-term) is already a ListLink, you only need to collect the result for each gene in a list and return the result as a ListLink. Hence we will have nested ListLinks in the result.

Comment (tanksha) [reply functions/gene_go_annotation.scm]: @Habush going through the results while bench marking, I am getting Go terms under namespace 'molecular_function' while I request for (list "cellular_component"  "biological_process")
Can you please check?

**Comment by tanksha** [comment]
@Habush going through the results while bench marking, I am getting Go terms under namespace 'molecular_function' while I request for (list "cellular_component"  "biological_process")
Can you please check?

File: functions/gene_go_annotation.scm (hunk 1 above)
Comment (Habush) [reply functions/gene_go_annotation.scm]: @tanksha Can you send me an example that reproduces an empty result? In my test, it returned a `ListLink` of the results.

Comment (tanksha) [reply functions/gene_go_annotation.scm]: It actually works, I don't know why it happened yesterday. But there is a serious issue that given any namespace, all three are being retrieved. You can even ask for empty string namespace and get all GO terms in three of the namespaces. I think it is the OrLink which is creating this problem, can you please check?

Comment (tanksha) [reply functions/gene_go_annotation.scm]: use the sample dataset [here](https://gist.github.com/tanksha/a2b8d6f6f771e5b3a0dc9fb8a97a3c9e) and try the following 
- (gene-go-annotation (list "SPAG9") (list "cellular_component"))
check the result, it also contains GO terms under molecular_function and bological_process

- (gene-go-annotation (list "SPAG9") (list ""))
You can even pass empty string as a namespace and the result contains all GO terms under three of the namespaces.

**Pull request merged**